blockchain/event_filter.go

```diff
@@ -0,0 +1,198 @@
+package blockchain
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db"
+)
+
+var errChunkSizeReached = errors.New("chunk size reached")
+
+type EventFilter struct {
+	txn             db.Transaction
+	fromBlock       uint64
+	toBlock         uint64
+	contractAddress *felt.Felt
+	keys            []*felt.Felt
+}
+
+type FilterRangeEnd byte
+
+const (
+	From FilterRangeEnd = iota
+	To
+)
+
+func newEventFilter(txn db.Transaction, contractAddress *felt.Felt, keys []*felt.Felt, fromBlock, toBlock uint64) *EventFilter {
+	return &EventFilter{
+		txn:             txn,
+		contractAddress: contractAddress,
+		keys:            keys,
+		fromBlock:       fromBlock,
+		toBlock:         toBlock,
+	}
+}
+
+// SetRangeEndBlockByNumber sets an end of the block range by block number
+func (f *EventFilter) SetRangeEndBlockByNumber(end FilterRangeEnd, blockNumber uint64) error {
+	_, err := blockHeaderByNumber(f.txn, blockNumber)
+	if err != nil {
+		return err
+	}
+	if end == From {
+		f.fromBlock = blockNumber
+	} else if end == To {
+		f.toBlock = blockNumber
+	} else {
+		return errors.New("undefined range end")
+	}
+	return nil
+}
+
+// SetRangeEndBlockByHash sets an end of the block range by block hash
+func (f *EventFilter) SetRangeEndBlockByHash(end FilterRangeEnd, blockHash *felt.Felt) error {
+	header, err := blockHeaderByHash(f.txn, blockHash)
+	if err != nil {
+		return err
+	}
+	return f.SetRangeEndBlockByNumber(end, header.Number)
+}
+
+// Close closes the underlying database transaction that provides the blockchain snapshot
+func (f *EventFilter) Close() error {
+	return f.txn.Discard()
+}
+
+type ContinuationToken struct {
+	fromBlock       uint64
+	processedEvents uint64
+}
+
+func (t *ContinuationToken) String() string {
+	return fmt.Sprintf("%d-%d", t.fromBlock, t.processedEvents)
+}
+
+func (t *ContinuationToken) FromString(str string) error {
+	_, err := fmt.Sscanf(str, "%d-%d", &t.fromBlock, &t.processedEvents)
+	return err
+}
```

Comment:
I like this simple approach to managing continuation tokens, I thought we may have to do something more complicated and maintain a record of different tokens similar to cookies.

Do you think parse checking is enough to return `ErrInvalidContinuationToken`?

There are no checks/tests to define what happens if a structurally correct token is passed but it is for a different request. The result could be unpredictable. Is it sufficient to assume that it is the user's responsibility to provide a correct token?

Reply:
> Do you think parse checking is enough to return ErrInvalidContinuationToken?

Yes I think it is. As long as we don't crash (we dont), it is acceptable. User shouldnt misuse the API and expect to get consistent results.